<center><h1><strong><font color="blue">Studi Kasus Data Wrangling</font></strong></h1></center>

![taudata Analytics](https://github.com/taudataid/eLearning/raw/master/images/covers/cover_ui-n-taudata.jpg "taudata Analytics")

In [ ]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

In [ ]:
# Code untuk Soal No 01 diberikan sebagai Contoh dan Bonus
# Pelajari dengan baik

nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN', right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_dosen = nilai_dosen.set_index(['NIDN']) 
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

nama  mean_nilai  std_nilai
NIDN                                      
123    Andri Basuki   85.750000   7.741355
234        zulfikar   80.428571  15.883504
345   Annisa Fitria   60.090909  16.220077
787            yudi   70.571429  13.010984

# Interpretasi 01: 

Setelah menghitung rata-rata dan standar deviasi nilai yang diberikan oleh dosen-dosen, terdapat dosen yang memberikan nilai cenderung lebih tinggi atau lebih rendah dibanding dosen lain, diantaranya sebagai berikut.

* Dosen yang memberikan nilai tertinggi secara rata-rata adalah Andri Basuki. Dapat dilihat dari rata-rata nilai yang diberikan Andri Basuki paling tinggi, yaitu 85.75. Selain itu, standar deviasinya paling rendah, yang mengindikasi nilai yang diberikan konsisten tinggi.

*   Disisi lain, dosen yang memberikan nilai kurang baik secara rata-rata adalah Annisa Fitria. Dapat dilihat dari rata-rata nilai yang diberikan Annisa Fitria paling rendah, yaitu 60.0909. Selain itu, standar deviasinya paling tinggi, yang mengindilkasikan nilai yang diberikan bervariasi. 

*   Dosen urutan kedua yang memberikan nilai tinggi adalah Zulfikar. Dapat dilihat rata-rata nilai yang diberikan Zulfikar adalah 80.428571.

# Interpretasi 02: 

* 

2) Menurut data yang ada, apakah terdapat perbedaan dalam pemberian nilai antara dosen dengan Pendidikan Terakhir S3 dan S2?

In [ ]:
pendidikan = pd.merge(dosen,nilai)
nilaipendidikan = pendidikan.groupby(pendidikan['Pendidikan_Terakhir']).mean()[['nilai']]
nilaipendidikan = nilaipendidikan.rename({'nilai':'mean_nilai'},axis=1)
sb = pendidikan.groupby(pendidikan['Pendidikan_Terakhir']).std()[['nilai']]
sb = sb.rename({'nilai':'std_nilai'},axis=1)
nilaipendidikan = pd.concat([nilaipendidikan, sb], axis=1)
nilaipendidikan

mean_nilai  std_nilai
Pendidikan_Terakhir                       
S2                    64.166667  15.561075
S3                    83.266667  12.068061

In [ ]:
# Code Untuk Soal No 02
pendidikan = pd.merge(dosen,nilai)
nilaipendidikan = pendidikan.groupby([pendidikan['Pendidikan_Terakhir'],pendidikan['nama']]).mean()[['nilai']]
nilaipendidikan = nilaipendidikan.rename({'nilai':'mean_nilai'},axis=1)
sb = pendidikan.groupby([pendidikan['Pendidikan_Terakhir'],pendidikan['nama']]).std()[['nilai']]
sb = sb.rename({'nilai':'std_nilai'},axis=1)
nilaipendidikan = pd.concat([nilaipendidikan, sb], axis=1)
nilaipendidikan

mean_nilai  std_nilai
Pendidikan_Terakhir nama                                
S2                  Annisa Fitria   60.090909  16.220077
                    yudi            70.571429  13.010984
S3                  Andri Basuki    85.750000   7.741355
                    zulfikar        80.428571  15.883504

dapat dilihat dari hasil yang keluar bahwa terdapat perbedaan hasil rata-rata antara S2 dan S3 dengan dosen 


*   dengan nilai rata-rata S2 = 64.166667 \
dosen Annisa Fitria = 60.090909\
dosen yudi          = 70.571429


*   sedangkan nilai rata-rata S3 = 83.266667\
dosen Andri Basuki = 85.750000\
dosen zulfikar = 80.428571




3. Apakah terdapat perbedaan performa akademik antara mahasiswa yang berasal dari Madrasah/SMK dan SMU?

In [ ]:
# Code Soal No 03
nilai_mahasiswa = nilai.groupby(nilai['nim']).mean()[['nilai']]
nilai_mahasiswa = nilai_mahasiswa.rename({'nilai':'mean_nilai'}, axis=1)
nilai_mahasiswa = pd.merge(mahasiswa, nilai_mahasiswa, left_on='nim', right_index=True)
nilai_mahasiswa = nilai_mahasiswa[['nim', 'Asal_Sekolah', 'mean_nilai']]
sb_mah = nilai.groupby(nilai['nim']).std()[['nilai']]
sb_mah = sb_mah.rename({'nilai':'std_nilai'}, axis=1)
nilai_mahasiswa = nilai_mahasiswa.set_index(['nim'])
nilai_mahasiswa = pd.concat([nilai_mahasiswa, sb_mah], axis=1)
nilai_mahasiswa.pivot_table(index='Asal_Sekolah')

mean_nilai  std_nilai
Asal_Sekolah                       
MA             71.937500  19.673045
SMK            85.900000   4.693409
SMU            66.416667  16.729164

# Interpretasi 03:

Ya, terdapat perbedaan performa akademik antara mahasiswa yang berasal dari Madrasah/SMK dan SMU. Diperoleh rata rata nilai mahasiswa asal Madrasah yaitu 71.9375 , rata rata nilai mahasiswa asal SMK 85.9000, dan rata rata nilai mahasiswa SMU 66.4167. Dari perolehan tersebut, diketahui bahwa performa akademik mahasiswa yang berasal dari SMK dan Madrasah lebih baik daripada SMU jika dilihat dari nilai mean nya.

Adapun diperoleh juga nilai standar deviasi. Diketahui bahwa nilai standar deviasi mahasiswa asal SMK lebih kecil dari pada SMU dan Madrasah. Nilai standar deviasi yang rendah menggambarkan persebaran nilai yang kecil yang mengimplikasikan bahwa nilai nilai mahasiswa SMK semakin akurat terhadap mean. Sedangkan jika nilai standar deviasi tinggi, maka dapat diketahui terdapat nilai yang beragam (adanya kesenjangan). Sehingga dari ketiga asal sekolah di atas, diperoleh bahwa mahasiswa yang berasal dari SMK memiliki performa akademik terbaik dibanding Madrasah dan SMU.

# Interpretasi 04:

* 

4. Apakah ada perbedaan yang nampak di data terhadap kemampuan akademik antara mahasiswa dan mahasiswi?

In [ ]:
# Code Soal No 04

#Untuk seluruh matkul
nil = pd.merge(nilai, mahasiswa)
nil = nil[['nim', 'id_mata_kuliah', 'nilai', 'Gender']]
mean_nil = nil.groupby(nil['Gender']).mean()[['nilai']]
mean_nil = mean_nil.rename({'nilai': 'mean_nilai'},axis=1)
sd_nil = nil.groupby(nil['Gender']).std()[['nilai']]
sd_nil = sd_nil.rename({'nilai': 'std_nilai'}, axis=1)
nilai_nil = pd.concat([mean_nil, sd_nil], axis=1)
nilai_nil

mean_nilai  std_nilai
Gender                       
Pria     73.888889  17.146047
Wanita   71.600000  17.112235

In [ ]:
#Untuk matkul KAL01
nilai_kal01 = nilai.set_index(['id_mata_kuliah'])
nilai_kal01 = nilai_kal01.drop(['CS01','KAL02','STA02'])
nilai_kal01 = pd.merge(nilai_kal01, mahasiswa)
nilai_kal01 = nilai_kal01[['nim', 'nilai','Gender']]
mean_kal01 = nilai_kal01.groupby(nilai_kal01['Gender']).mean()[['nilai']]
mean_kal01 = mean_kal01.rename({'nilai': 'mean_KAL01'},axis=1)
sd_kal01 = nilai_kal01.groupby(nilai_kal01['Gender']).std()[['nilai']]
sd_kal01 = sd_kal01.rename({'nilai': 'std_KAL01'},axis=1)
nilai_kal01 = pd.concat([mean_kal01, sd_kal01], axis=1)
#Untuk matkul CS01
nilai_cs01 = nilai.set_index(['id_mata_kuliah'])
nilai_cs01 = nilai_cs01.drop(['KAL01','KAL02','STA02'])
nilai_cs01 = pd.merge(nilai_cs01, mahasiswa)
nilai_cs01 = nilai_cs01[['nim', 'nilai','Gender']]
mean_cs01 = nilai_cs01.groupby(nilai_cs01['Gender']).mean()[['nilai']]
mean_cs01 = mean_cs01.rename({'nilai': 'mean_CS01'},axis=1)
sd_cs01 = nilai_cs01.groupby(nilai_cs01['Gender']).std()[['nilai']]
sd_cs01 = sd_cs01.rename({'nilai': 'std_CS01'},axis=1)
nilai_cs01 = pd.concat([mean_cs01, sd_cs01], axis=1)
#Untuk matkul STA02
nilai_sta02 = nilai.set_index(['id_mata_kuliah'])
nilai_sta02 = nilai_sta02.drop(['CS01','KAL01','KAL02'])
nilai_sta02 = pd.merge(nilai_sta02, mahasiswa)
nilai_sta02 = nilai_sta02[['nim', 'nilai','Gender']]
mean_sta02 = nilai_sta02.groupby(nilai_sta02['Gender']).mean()[['nilai']]
mean_sta02 = mean_sta02.rename({'nilai': 'mean_STA02'},axis=1)
sd_sta02 = nilai_sta02.groupby(nilai_sta02['Gender']).std()[['nilai']]
sd_sta02 = sd_sta02.rename({'nilai': 'std_STA02'},axis=1)
nilai_sta02 = pd.concat([mean_sta02, sd_sta02], axis=1)
#Untuk matkul KAL02
nilai_kal02 = nilai.set_index(['id_mata_kuliah'])
nilai_kal02 = nilai_kal02.drop(['CS01','KAL01','STA02'])
nilai_kal02 = pd.merge(nilai_kal02, mahasiswa)
nilai_kal02 = nilai_kal02[['nim', 'nilai','Gender']]
mean_kal02 = nilai_kal02.groupby(nilai_kal02['Gender']).mean()[['nilai']]
mean_kal02 = mean_kal02.rename({'nilai': 'mean_KAL02'},axis=1)
sd_kal02 = nilai_kal02.groupby(nilai_kal02['Gender']).std()[['nilai']]
sd_kal02 = sd_kal02.rename({'nilai': 'std_KAL02'},axis=1)
nilai_kal02 = pd.concat([mean_kal02, sd_kal02], axis=1)
nilai_matkul = pd.concat([nilai_kal01, nilai_cs01, nilai_sta02, nilai_kal02], axis=1)
nilai_matkul

mean_KAL01  std_KAL01  mean_CS01   std_CS01  mean_STA02  std_STA02  \
Gender                                                                       
Pria     81.222222  10.986102       91.0   1.000000   48.666667   8.082904   
Wanita   74.833333  15.380724       72.5  17.559423   65.250000  24.212600   

        mean_KAL02  std_KAL02  
Gender                         
Pria          60.0        5.0  
Wanita        74.0        NaN

Berdasarkan rata-rata nilai yang diperoleh, terdapat perbedaan nilai, dengan nilai mahasiswa/pria lebih tinggi pada 73.89 dibandingkan nilai mahasiswi/wanita pada 71.6. 
Pada rata-rata nilai yang diperoleh berdasarkan mata kuliah, perbedaan nilai rata-rata antara mahasiswa dan mahasiswi yang diperoleh lebih terlihat. 

*   Pada Kalkulus Dasar, nilai mahasiswa 81.2 sedangkan nilai mahasiswi 74.83
*   Pada Pendahuluan Ilmu Komputer, nilai mahasiswa 91.0 dan nilai mahasiswi 72.5
*   Pada Statistika Matematika, nilai mahasiswa 48.67 dan nilai mahasiswi 65.25
*   Pada Kalkulus Lanjut, nilai mahasiswa 60 dan nilai mahasiswi 74

Dilihat dari rata-rata nilai, dapat disimpulkan bahwa ada perbedaan kemampuan akademik antara mahasiswa dan mahasiswi.

# Interpretasi 05

Apakah ada perbedaan prestasi akademik yang nampak di data antar angkatan mahasiswa?

dapat kita lihat bahwa terdapat perbedaan prestasi akademik antar angaktan mahasiswa dilihat dari mean(rata-tara)nilai antar angkatan dan standar deviasi nilai dari tiap angkatan


In [ ]:
#code soal no 5
prestasi_angkatan = nilai.groupby(nilai['nim']).mean()[['nilai']]
prestasi_angkatan = prestasi_angkatan.rename({'nilai': 'mean_nilai'}, axis=1)
prestasi_angkatan = pd.merge(mahasiswa, prestasi_angkatan, left_on='nim', right_index=True)
prestasi_angkatan = prestasi_angkatan[['nim', 'Angkatan', 'mean_nilai']]
sb = nilai.groupby(nilai['nim']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
prestasi_angkatan = prestasi_angkatan.set_index(['nim']) 
prestasi_angkatan = pd.concat([prestasi_angkatan, sb], axis=1)
prestasi_angkatan.pivot_table(index='Angkatan')

mean_nilai  std_nilai
Angkatan                       
2019        75.59375  13.385601
2020        73.00000  13.547877

# Interpretasi 06

* 

In [ ]:
# Code soal No 06

# Interpretasi 06

*  